args/audio/systems/audiosystem.inl

```diff
@@ -0,0 +1,279 @@
+		ALCint srate;
+		alcGetIntegerv(data::alDevice, ALC_FREQUENCY, 1, &srate);
+		log::info("OpenAL info:\n\n\t\t\tOpenAL information\n\tVendor: {}\n\tVersion: {}\n\tRenderer: {}\n\tOpenAl Extensions: {}\n\tALC Extensions: {}\n\tDevice samplerate: {}\n",
+			vendor, version, renderer, openALExtensions, ALCExtensions, srate);
+	}
+
+	inline void AudioSystem::update(time::span deltatime)
+	{
+		alcMakeContextCurrent(data::alContext);
+
+		int i;
+
+		for (auto entity : sourceQuery)
+		{
+			auto sourceHandle = entity.get_component_handle<audio_source>();
+
+			audio_source source = sourceHandle.read();
+			position p = entity.read_component<position>();
+			position& previousP = m_sourcePositions.at(sourceHandle);
+			math::vec3 vel = previousP - p;
+			previousP = p;
+			alSource3f(source.m_sourceId, AL_POSITION, p.x, p.y, p.z);
+			alSource3f(source.m_sourceId, AL_VELOCITY, vel.x, vel.y, vel.z);
+			log::debug(source.m_changes);
+			if (source.m_changes & 1)
+			{
+				// Pitch has changed
+				alSourcef(source.m_sourceId, AL_PITCH, source.getPitch());
+			}
+			if (source.m_changes & 2)
+			{
+				// Gain has changed
+				alSourcef(source.m_sourceId, AL_GAIN, source.getGain());
+			}
+
+			source.clearChanges();
+			sourceHandle.write(source);
+		}
+
+		if (m_listenerEnt)
+		{
+			position p = m_listenerEnt.read_component<position>();
+			rotation r = m_listenerEnt.read_component<rotation>();
+
+			setListener(p, r);
+
+			math::vec3 vel = m_listenerPosition - p;
+			m_listenerPosition = p;
+			alListener3f(AL_VELOCITY, vel.x, vel.y, vel.z);
+		}
+
+		alcMakeContextCurrent(nullptr);
+	}
+
+#pragma region Component creation&destruction
+	inline void AudioSystem::onAudioSourceComponentCreate(events::component_creation<audio_source>* event)
+	{
+		alcMakeContextCurrent(data::alContext);
+
+		auto handle = event->entity.get_component_handle<audio_source>();
+		audio_source a = handle.read();
+
+		// do something with a.
+		if (!initSource(a))
+		{
+			handle.destroy();
+
+			alcMakeContextCurrent(nullptr);
+
+#if defined(AUDIO_EXIT_ON_FAIL)
+			raiseEvent<events::exit>();
+#endif
+			return;
+		}
+
+		// NOTE TO SELF:
+		// REMOVE THE AUTO PLAY (TESTING PURPOSE)
+		log::debug("playing sound");
+		alSourcePlay(a.m_sourceId);
+
+		m_sourcePositions.emplace(handle, event->entity.read_component<position>());
+
+		handle.write(a);
+		++data::sourceCount;
+
+		alcMakeContextCurrent(nullptr);
+	}
+
+
+	inline void AudioSystem::onAudioSourceComponentDestroy(events::component_destruction<audio_source>* event)
+	{
+		alcMakeContextCurrent(data::alContext);
+
+		auto handle = event->entity.get_component_handle<audio_source>();
+		audio_source a = handle.read();
+
+		m_sourcePositions.erase(handle);
+
+		handle.write(a);
+		--data::sourceCount;
+
+		alcMakeContextCurrent(nullptr);
+	}
+
+	inline void AudioSystem::onAudioListenerComponentCreate(events::component_creation<audio_listener>* event)
+	{
+		alcMakeContextCurrent(data::alContext);
```

Comment:
what if making the context current fails because another thread is utilizing the context?